Script zfs-functions missing from generated .deb packages, causes kernel panic

### System information
<!--  add version after "|" character -->
Type | Version/Name
 --- | --- 
Distribution Name	| Ubuntu Server
Distribution Version	| 18.04.1
Linux Kernel	| 4.15.0-34-generic
Architecture	| amd64
ZFS Version	| 0.8.0-rc1
<!-- 
Commands to find ZFS/SPL versions:
modinfo zfs | grep -iw version
modinfo spl | grep -iw version 
-->

### Describe the problem you're observing

When building DEB packages of ZFSonLinux with 'make deb', the script 'zfs-functions' does not end up in any of the generated DEB files.

At the very least, this results in a kernel panic when attempting to boot, if you are using a ZFS root filesystem. I would imagine it probably breaks other stuff too, but have not tested this.

The file does get generated in the source tree at ./etc/init.d/zfs-functions. Manually copying this to /etc/zfs/ (which is where Ubuntu expects it) allows the system to boot and function normally.

Modifying the DEB file to manually place this file in /etc/zfs/ also results in a package that installs and works correctly.

### Describe how to reproduce the problem

Compile ZFS, then run 'make deb'. Attempt to use the generated packages.

### Include any warning/errors/backtraces from the system logs
<!-- 
*IMPORTANT* - Please mark logs and text output from terminal commands 
or else Github will not display them correctly. 
An example is provided below.

Example:
```
this is an example how log text should be marked (wrap it with ```)
```
-->

N/A

The full log is quite large, but here's the relevant part:

```
[ -x ./scripts/zfs-tests.sh ] && \
    ./scripts/zfs-tests.sh -c
make[5]: Entering directory '/tmp/zfs-build-acingram-dl7bLSTg/BUILD/zfs-0.8.0'
make[5]: Nothing to be done for 'install-exec-am'.
make  install-data-hook
make[6]: Entering directory '/tmp/zfs-build-acingram-dl7bLSTg/BUILD/zfs-0.8.0'
make[6]: Nothing to be done for 'install-data-hook'.
make[6]: Leaving directory '/tmp/zfs-build-acingram-dl7bLSTg/BUILD/zfs-0.8.0'
make[5]: Leaving directory '/tmp/zfs-build-acingram-dl7bLSTg/BUILD/zfs-0.8.0'
make[4]: Leaving directory '/tmp/zfs-build-acingram-dl7bLSTg/BUILD/zfs-0.8.0'
make[3]: Leaving directory '/tmp/zfs-build-acingram-dl7bLSTg/BUILD/zfs-0.8.0'
make[2]: Leaving directory '/tmp/zfs-build-acingram-dl7bLSTg/BUILD/zfs-0.8.0'
+ find /tmp/zfs-build-acingram-dl7bLSTg/BUILDROOT/zfs-0.8.0-1.x86_64/usr/local/lib -name *.la -exec rm -f {} ;
+ /usr/lib/rpm/brp-compress
+ /usr/lib/rpm/brp-strip /usr/bin/strip
+ /usr/lib/rpm/brp-strip-static-archive /usr/bin/strip
+ /usr/lib/rpm/brp-strip-comment-note /usr/bin/strip /usr/bin/objdump
Processing files: zfs-0.8.0-1.x86_64
error: File not found: /tmp/zfs-build-acingram-dl7bLSTg/BUILDROOT/zfs-0.8.0-1.x86_64/usr/local/etc/zfs/zfs-functions
error: File not found: /tmp/zfs-build-acingram-dl7bLSTg/BUILDROOT/zfs-0.8.0-1.x86_64/etc/default/zfs


RPM build errors:
    File not found: /tmp/zfs-build-acingram-dl7bLSTg/BUILDROOT/zfs-0.8.0-1.x86_64/usr/local/etc/zfs/zfs-functions
    File not found: /tmp/zfs-build-acingram-dl7bLSTg/BUILDROOT/zfs-0.8.0-1.x86_64/etc/default/zfs
Makefile:1335: recipe for target 'rpm-common' failed
make[1]: *** [rpm-common] Error 1
make[1]: Leaving directory '/home/acingram/zfs-test/zfs'
Makefile:1294: recipe for target 'rpm-utils' failed
make: *** [rpm-utils] Error 2
```

--- a/contrib/initramfs/Makefile.am
+++ b/contrib/initramfs/Makefile.am
@@ -6,9 +6,14 @@ initrd_SCRIPTS = \
 SUBDIRS = hooks scripts
 
 EXTRA_DIST = \
+	$(top_srcdir)/etc/init.d/zfs \
+	$(top_srcdir)/etc/init.d/zfs-functions \
 	$(top_srcdir)/contrib/initramfs/conf.d/zfs \
 	$(top_srcdir)/contrib/initramfs/conf-hooks.d/zfs \
 	$(top_srcdir)/contrib/initramfs/README.initramfs.markdown
+
+$(top_srcdir)/etc/init.d/zfs $(top_srcdir)/etc/init.d/zfs-functions:
+	$(MAKE) -C $(top_srcdir)/etc/init.d zfs zfs-functions
 
 install-initrdSCRIPTS: $(EXTRA_DIST)
 	for d in conf.d conf-hooks.d scripts/local-top; do \
@@ -21,8 +26,9 @@ install-initrdSCRIPTS: $(EXTRA_DIST)
 		cp $(top_builddir)/contrib/initramfs/$$d/zfs \
 		    $(DESTDIR)$(initrddir)/$$d/; \
 	done
-	if [ -f $(top_builddir)/etc/init.d/zfs ]; then \
-		$(MKDIR_P) $(DESTDIR)$(DEFAULT_INITCONF_DIR); \
-		cp $(top_builddir)/etc/init.d/zfs \
-		    $(DESTDIR)$(DEFAULT_INITCONF_DIR)/; \
-	fi
+	$(MKDIR_P) $(DESTDIR)$(DEFAULT_INITCONF_DIR); \
+	cp $(top_builddir)/etc/init.d/zfs \
+	    $(DESTDIR)$(DEFAULT_INITCONF_DIR)/; \
+	$(MKDIR_P) $(DESTDIR)$(sysconfdir)/zfs; \
+	cp $(top_builddir)/etc/init.d/zfs-functions \
+	    $(DESTDIR)$(sysconfdir)/zfs/

--- a/contrib/initramfs/hooks/zfs.in
+++ b/contrib/initramfs/hooks/zfs.in
@@ -4,16 +4,18 @@
 #
 
 # This hook installs udev rules for ZoL.
-PREREQ="zdev"
+PREREQ="udev"
 
 # These prerequisites are provided by the zfsutils package. The zdb utility is
 # not strictly required, but it can be useful at the initramfs recovery prompt.
 COPY_EXEC_LIST="@sbindir@/zdb @sbindir@/zpool @sbindir@/zfs"
 COPY_EXEC_LIST="$COPY_EXEC_LIST @mounthelperdir@/mount.zfs @udevdir@/vdev_id"
+COPY_EXEC_LIST="$COPY_EXEC_LIST @udevdir@/zvol_id"
 COPY_FILE_LIST="/etc/hostid @sysconfdir@/zfs/zpool.cache"
-COPY_FILE_LIST="$COPY_FILE_LIST @sysconfdir@/default/zfs"
+COPY_FILE_LIST="$COPY_FILE_LIST @DEFAULT_INITCONF_DIR@/zfs"
 COPY_FILE_LIST="$COPY_FILE_LIST @sysconfdir@/zfs/zfs-functions"
 COPY_FILE_LIST="$COPY_FILE_LIST @sysconfdir@/zfs/vdev_id.conf"
+COPY_FILE_LIST="$COPY_FILE_LIST @udevruledir@/60-zvol.rules"
 COPY_FILE_LIST="$COPY_FILE_LIST @udevruledir@/69-vdev.rules"
 
 # These prerequisites are provided by the base system.

--- a/etc/systemd/system/Makefile.am
+++ b/etc/systemd/system/Makefile.am
@@ -31,5 +31,9 @@ $(systemdunit_DATA) $(systemdpreset_DATA):%:%.in
 		-e 's,@sysconfdir\@,$(sysconfdir),g' \
 		$< >'$@'
 
+install-data-hook:
+	$(MKDIR_P) "$(DESTDIR)$(systemdunitdir)"
+	ln -s /dev/null "$(DESTDIR)$(systemdunitdir)/zfs-import.service"
+
 distclean-local::
 	-$(RM) $(systemdunit_DATA) $(systemdpreset_DATA)

--- a/rpm/generic/zfs.spec.in
+++ b/rpm/generic/zfs.spec.in
@@ -433,6 +433,14 @@ systemctl --system daemon-reload >/dev/null || true
 %{_udevdir}/vdev_id
 %{_udevdir}/zvol_id
 %{_udevdir}/rules.d/*
+%if ! 0%{?_systemd} || 0%{?_initramfs}
+# Files needed for sysvinit and initramfs-tools
+%{_sysconfdir}/%{name}/zfs-functions
+%config(noreplace) %{_initconfdir}/zfs
+%else
+%exclude %{_sysconfdir}/%{name}/zfs-functions
+%exclude %{_initconfdir}/zfs
+%endif
 %if 0%{?_systemd}
 %{_unitdir}/*
 %{_presetdir}/*
@@ -440,9 +448,10 @@ systemctl --system daemon-reload >/dev/null || true
 %{_systemdgeneratordir}/*
 %else
 %config(noreplace) %{_sysconfdir}/init.d/*
-%config(noreplace) %{_initconfdir}/zfs
 %endif
-%config(noreplace) %{_sysconfdir}/%{name}
+%config(noreplace) %{_sysconfdir}/%{name}/zed.d/*
+%config(noreplace) %{_sysconfdir}/%{name}/zpool.d/*
+%config(noreplace) %{_sysconfdir}/%{name}/vdev_id.conf.*.example
 %attr(440, root, root) %config(noreplace) %{_sysconfdir}/sudoers.d/*
 
 %files -n libzpool2

